src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+/// be `true` if the item that `used_generic_parameters` was invoked on is a closure.
+fn mark_used_by_default_parameters<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    leaf_is_closure: bool,
+    used_parameters: &mut BitSet<u32>,
+) {
+    for param in &generics.params {
+        debug!("mark_used_by_default_parameters: param={:?}", param);
+        // Consider any of the dummy generic parameters that closures have as used.
+        if leaf_is_closure && param.def_id == def_id {
+            used_parameters.insert(param.index);
+        }
+
+        // Consider every non-generic parameter used by default.
+        match param.kind {
+            ty::GenericParamDefKind::Type { .. } => (),
+            _ => {
+                used_parameters.insert(param.index);
+            }
+        }
+    }
+
+    if let Some(parent) = generics.parent {
+        mark_used_by_default_parameters(
+            tcx,
+            def_id,
+            tcx.generics_of(parent),
+            false,
+            used_parameters,
+        );
+    }
+}
+
+/// Emit an error for the function represented by `def_id`, labelling each generic parameter which
+/// was unused.
+fn emit_used_generic_parameters_error<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    generics: &'tcx ty::Generics,
+    used_parameters: &BitSet<u32>,
+) {
+    debug!("emit_used_generic_parameters_error: used_parameters={:?}", used_parameters);
+    let fn_span = match tcx.opt_item_name(def_id) {
+        Some(ident) => ident.span,
+        _ => tcx.def_span(def_id),
+    };
+
+    let mut err = tcx.sess.struct_span_err(fn_span, "item has unused generic parameters");
+
+    let mut unused_parameters = BitSet::new_filled(used_parameters.domain_size());
+    unused_parameters.subtract(used_parameters);
+    debug!("emit_used_generic_parameters_error: unused_parameters={:?}", unused_parameters);
+
+    for index in unused_parameters.iter() {
+        let param = ty::ParamTy::new(index, Symbol::intern(&format!("P{}", index)));
+        // Can safely call `type_param` as all non-type parameters are considered used.
+        let generic_param_def = generics.type_param(&param, tcx);
+        debug!(
+            "emit_used_generic_parameters_error: generic_param_def={:?} param={:?}",
+            generic_param_def, param
+        );
+        let def_span = tcx.def_span(generic_param_def.def_id);
+        err.span_label(
+            def_span,
+            &format!("generic parameter `{}` is unused", generic_param_def.name),
+        );
+    }
+
+    err.emit();
+}
+
+/// Visitor used to aggregate generic parameter uses.
+struct UsedGenericParametersVisitor<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    body: ReadOnlyBodyAndCache<'a, 'tcx>,
+    used_parameters: &'a mut BitSet<u32>,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, ty: T) {
```

Comment:
`value: T` instead of `ty: T` (this could be anything).

Reply:
I've fixed this in dd24b5b.